Add EIP-4910: Royalty Bearing NFTs (#4910)
ethereum/EIPs: EIPS/eip-4910.md

---
eip: 4910
title: Royalty Bearing NFTs
description: Extension of the ERC-721 standard for NFTs to correctly define, process, and pay (hierarchical) onchain royalties from NFT sales, going beyond EIP-2981.
author: Andreas Freund (@Therecanbeonlyone1969) 
discussions-to: https://ethereum-magicians.org/t/royalty-bearing-nfts/8453
status: Draft
type: Standards Track
category: ERC
created: 2022-03-14

> requires: 721, 165, 2771

Author: Removed reference to 2771 s a requirement, and merely referred to it as a non-normative example to represent `msg.sender`

---

## Abstract

> The proposal directly connects NFTs and royalties in a smart contract architecture extending the ERC721 standard, with the aim of precluding central authorities from manipulating or circumventing payments to those who are legally entitled to them.

Reviewer: ```suggestion
The proposal directly connects NFTs and royalties in a smart contract architecture extending the [ERC-721](./eip-721.md) standard, with the aim of precluding central authorities from manipulating or circumventing payments to those who are legally entitled to them.
```
Please use the proper reference notation as described in EIP-1.

#### Sending a Payout Payment

The following steps need to be taken:
* find the RA Sub Account based on `RAaccount` and the `subaccountPos` and extract the balance
* extract `tokentype` from the Sub Account
* based on the token type, send the payout payment (not exceeding the available balance)

## Specification
The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.

### Data Structures: Specifications and Rationales

#### Royalty Account and Royalty Sub Accounts

In order to create an interconnected data structure linking NFTs to RAs that is search optimized requires to make the following additions to the global data structures of an ERC721.

Note, a Royalty Account is defined as a collection of Royalty Sub Accounts linked to a meta account. This meta account is comprised of general account identifiers particular to the NFT it is linked to such as asset identifier, parent identifier etc.

> #### **[R1]**

Reviewer: This header format makes the EIP rather verbose, you may consider a different format.

Author: @lightclient this allows each requirement to be individually referenced with links which makes discussions etc. much easier. See also the Baseline Protocol Standard as another standard using this approach to make requirements referenceable ... https://github.com/eea-oasis/baseline-standard/blob/main/core/baseline-core-v1.0-psd01.md.

Would prefer to leave it this way, please.

Author: resolved it using `<a name="tag_name"></a>`

```
#### **[R61]**

*The internal `_royaltyPayOut` function MUST perform the following actions:
* *send the payment to the `payoutaccount`.*
* *update the `royaltybalance` of the `RAsubaccount` of the Royalty Account upon successful transfer.*

#### Sending a Payout Payment

#### **[R62]**

*The following steps MUST be taken to send out a royalty payment to its recipient:*
* *find the Royalty Sub Account.*
* *extract `tokentype` from the Royalty Sub Account.*
* *based on the token type send to the `payoutAccount` either*
    * *'ETH' / relevant protocol token or*
    * *another token based on token type* 
* *and only if the payout transaction is successful, deduct `amount` from `royaltybalance` of the Royalty Sub Account,`RAsubaccount`, and then return `true` as the function return parameter, otherwise return `false`.* 

## Backwards Compatibility
This EIP is backwards compatible to the ERC721 standard introducing new interfaces and functionality but retaining the core interfaces and functionality of the ERC-721 standard.

## Test Cases
A full test suite is part of the reference implementation.

## Reference Implementation

> The Treetrunk reference implementation of the standard can be found [here](https://github.com/treetrunkio/treetrunk-nft-reference-implementation).

Reviewer: The reference implementation should be included in the `assets` folder. See EIP-1 for details.

Author: done

Author: could not add the entire repo as a subfolder. Opted to add an md file with a link to the reference implementation. @lightclient

Reviewer 2: You should not be including external links anywhere in the EIP.  EIPs should stand on their own without external dependencies.  The Reference Implementation is optional, so you can just leave it out if it doesn't make sense to have a reference implementation.

Generally speaking, if the EIP is too complex to write a simple reference implementation (it doesn't have to be production worthy, it should focus on readability) that suggests the standard may be too complex.  This isn't always true, but it is a reasonable hint.

Author: @MicahZoltu I got an error on commit when adding a link to a folder under `../assets/eip-4910`. I will simply add the reference implementation folder and link to the eip-4910 folder under assets in the md file. The reference implementation is very instructive for other implementers and contains full test case coverage with extensive tooling and security tests that will help implementers.

Hence, it is not a question of size or complexity of the reference implementation but rather how to refer to it.

---
eip: 4910
title: Royalty Bearing NFTs
description: Extension of the ERC-721 standard for NFTs to correctly define, process, and pay (hierarchical) onchain royalties from NFT sales, going beyond EIP-2981.
author: Andreas Freund (@Therecanbeonlyone1969) 
discussions-to: https://ethereum-magicians.org/t/royalty-bearing-nfts/8453
status: Draft
type: Standards Track
category: ERC
created: 2022-03-14
requires: 721, 165
---

## Abstract
The proposal directly connects NFTs and royalties in a smart contract architecture extending the [ERC-721](./eip-721.md) standard, with the aim of precluding central authorities from manipulating or circumventing payments to those who are legally entitled to them.

> The proposal builds upon the [OpenZeppelin Smart Contract Toolbox](https://github.com/OpenZeppelin/openzeppelin-contracts) architecture, and extends it to include royalty account management (CRUD), royalty balance and payments management, simple trading capabilities -- Listing/De-Listing/Buying -- and capabilities to trace trading on exchanges. The royalty management capabilities allow for hierarchical royalty structures, referred to herein as royalty trees, to be established by logically connecting a "parent" NFT to its "children", and recursively enabling NFT "children" to have more children. 

Reviewer 3: Based on the... heft of the reference implementation, and this link to OZ, I have to ask: is it likely that there will be multiple implementations of this standard? If not, would this be better as a PR into OZ itself?

Author: @SamWilsn there will be more than one implementation of this standard. I think it will eventually make it into OZ via a PR though.

> The proposal builds upon the [OpenZeppelin Smart Contract Toolbox](https://github.com/OpenZeppelin/openzeppelin-contracts) architecture, and extends it to include royalty account management (CRUD), royalty balance and payments management, simple trading capabilities -- Listing/De-Listing/Buying -- and capabilities to trace trading on exchanges. The royalty management capabilities allow for hierarchical royalty structures, referred to herein as royalty trees, to be established by logically connecting a "parent" NFT to its "children", and recursively enabling NFT "children" to have more children. 

Reviewer 3: No external links please!

```suggestion
The proposal builds upon the OpenZeppelin Smart Contract Toolbox architecture, and extends it to include royalty account management (CRUD), royalty balance and payments management, simple trading capabilities -- Listing/De-Listing/Buying -- and capabilities to trace trading on exchanges. The royalty management capabilities allow for hierarchical royalty structures, referred to herein as royalty trees, to be established by logically connecting a "parent" NFT to its "children", and recursively enabling NFT "children" to have more children. 
```

Author: thank you for the reminder. Done.

In the example below, the artist who created the original is eligible to receive proceeds from every sale, and resale, of a print. 

![Fig1](/assets/eip-4910/eip-4910-print-families.png)

The basic concept for hierarchical royalties utilizing the above "ancestry concept" is demonstrated in the figure below.

![Fig2](/assets/eip-4910/eip-4910-royalties.png)

In order to solve for the complicated inheritance problem, this proposal breaks down the recursive problem of the hierarchy tree of depth N into N separate problems, one for each layer. This allows us to traverse the tree from its lowest level upwards to its root most efficiently.

This affords creators, and the distributors of art derived from the original, the opportunity to achieve passive income from the creative process, enhancing the value of an NFT, since it now not only has intrinsic value but also comes with an attached cash flow.

> ## Specification Outline

Reviewer 3: Not a valid section header as per EIP-1

```suggestion
## Specification

### Outline
```

Then bump the other sections in by one level.

Author: ok. thank you

Author: done

* Adding structs for a Royalty Account and associated Royalty Sub Accounts to establish the concept of a Royalty Account with sub accounts.
* Defining an `raAccountId` as the keccak256 hash of `tokenId`, the actual `owner` address, and the current block number, `block.blocknumber`
* Mapping a `tokenId` to an `raAccountID` in order to connect an RA `raAccountId` to a `tokenId`
* Mapping the `raAccountID` to a `RoyaltyAccount` in order to connect the account identifier to the actual account.
* An `ancestry` mapping of the parent-to-child NFT relationship
* A mapping of supported token types to their origin contracts and last validated balance (for trading and royalty payment purposes)
* A mapping with a struct for a registered payment to be made in the `executePayment` function and validated in `safeTransferFrom`. This is sufficient, because a payment once received and distributed in the `safeTransferFrom` function will be removed from the mapping.
* A mapping for listing NFTs to be sold

### Royalty Account Functions

> Definitions and interfaces for the Royalty Account RUD (Read-Update-Delete) functions. Because the RA is created in the minting function, there is no need to have a function to create a royalty account separately.

Reviewer 3: Is this section missing the actual "definitions and interfaces" or are the other headers at the wrong level?

Author: corrected ... I hope

> ### Minting of a royalty bearing NFT

Reviewer 3: I'd go through and double check the capitalization of your headers.

Author: done and corrected.

There are two versions of the payout function -- a `public` function and an `internal` function.

The public function has the following interface:
```
function royaltyPayOut (uint256 tokenId, address RAsubaccount, address payable payoutAccount, uint256 amount) public virtual nonReentrant returns (bool)
```

where we only need the `tokenId`, the RA Sub Account address, `_RAsubaccount` which is the `owner`, and the amount to be paid out, `_amount`. Note that the function has [`nonReentrant` modifier protection](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/security/ReentrancyGuard.sol), because funds are being payed out.

#### Sending a Payout Payment

The following steps need to be taken:
* find the RA Sub Account based on `RAaccount` and the `subaccountPos` and extract the balance
* extract `tokentype` from the Sub Account
* based on the token type, send the payout payment (not exceeding the available balance)

## Specification
The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.

> ### Data Structures: Specifications and Rationales

Reviewer 3: Rationale should be in the rationale section.

Author: did so.

* A mapping for listing NFTs to be sold

#### Royalty Account Functions

Definitions and interfaces for the Royalty Account RUD (Read-Update-Delete) functions. Because the RA is created in the minting function, there is no need to have a function to create a royalty account separately.

#### Minting of a Royalty Bearing NFT

When an NFT is minted, an RA must be created and associated with the NFT and the NFT owner, and, if there is an ancestor, with the ancestor's RA. To this end the specification utilizes the `_safemint` function in a newly defined `mint` function and applies various business rules on the input variables.

#### Listing NFTs for Sale and removing a Listing

Authorized user addresses can list NFTs for sale for non-exchange mediated NFT purchases.

#### Payment Function from Buyer to Seller

To avoid royalty circumvention, a buyer will always pay the NFT contract directly and not the seller. The seller is paid through the royalty distribution and can later request a payout.

The payment process depends on whether the payment is received in ETH or an ERC 20 token:

> * ERC 20 Token

Reviewer 3: Throughout:

```suggestion
* ERC-20 Token
```

This is the final part of the proposal.

There are two versions of the payout function -- a `public` function and an `internal` function.

The public function has the following interface:
```
function royaltyPayOut (uint256 tokenId, address RAsubaccount, address payable payoutAccount, uint256 amount) public virtual nonReentrant returns (bool)
```

where we only need the `tokenId`, the RA Sub Account address, `_RAsubaccount` which is the `owner`, and the amount to be paid out, `_amount`. Note that the function has `nonReentrant` modifier protection, because funds are being payed out.

#### Sending a Payout Payment

The following steps need to be taken:
* find the RA Sub Account based on `RAaccount` and the `subaccountPos` and extract the balance
* extract `tokentype` from the Sub Account
* based on the token type, send the payout payment (not exceeding the available balance)

### Data Structures

#### Rationale: Royalty Account and Royalty Sub Accounts

In order to create an interconnected data structure linking NFTs to RAs that is search optimized requires to make the following additions to the global data structures of an ERC721.

> Note, a Royalty Account is defined as a collection of Royalty Sub Accounts linked to a meta account. This meta account is comprised of general account identifiers particular to the NFT it is linked to such as asset identifier, parent identifier etc.

Reviewer 3: All the "Rationale" has to be under the top-level Rationale section, not in the "Specification" section. The "Specification" section should contain the minimal amount of text required for an implementer to implement the spec.

That said, I think these mini-rationale sections are small enough that you could do them inline, like this:

```markdown
### Data Structures

#### Royalty Account and Royalty Sub Accounts

In order to create an interconnected data structure linking NFTs to RAs that is search optimized requires to make the following additions to the global data structures of an ERC721.

Note, a Royalty Account is defined as a collection of Royalty Sub Accounts linked to a meta account. This meta account is comprised of general account identifiers particular to the NFT it is linked to such as asset identifier, parent identifier etc.

##### **[R1]**

*There MUST be a struct for a Royalty Account as minimally defined below.*
`` `
///
/// @dev Definition of the struct for a Royalty Account.
///
/// @param  assetId is the tokenId of the NFT to which the RA belongs
/// @param  parent is the tokenId of the NFT from which the NFT to which this RA belongs, is derived
/// @param  royaltysplitForItsChildren is the royalty percentage to be paid to the RA belonging to the parent NFT from its immediate descendants
/// @param  tokentype of the royalty balance in this Royalty Account
/// @param  balance is the total RA account balance and must be equal to the sum of the subaccount balances
/// @param  the struct array for the Royalty Sub-Accounts
struct RoyaltyAccount {
    uint256 assetId;
    uint256 parent;
    uint256 royaltySplitForItsChildren;
    string tokentype;
    uint256 balance;
    RASubAccount[] rasubaccount;
}
`` `
```

Sorry I wasn't more clear in my initial review ^^;;

Author: @SamWilsn no issue. fixed it.

Author: all good. did that,

#### Specification: Royalty Account and Royalty Sub Accounts

> ##### **[R1]**

Reviewer 3: I haven't tried it myself, but would footnotes (`[^1]`) work for these? Failing that, maybe `<a id="r1">...</a>` is supported?

Author: will give it a whirl and see if it works @SamWilsn

Author: done ... using <a name="name_tag">...</a> works

mapping (uint256 => address) private _tokenindextoRA;
```

##### **[R5]**

*`raAccountID` MUST be mapped to a `RoyaltyAccount` in order to connect the Royalty  Account identifier to the Royalty Account.*
```
mapping (address => RoyaltyAccount) private _royaltyaccount;
```

#### Rationale: Print (Child) NFTs

The set of requirement to manage Parent-Child NFT Relationships and constraints at each level of the NFT (family) tree e.g. number of children permitted, NFT parents have to be mapped to their immediate NFT children are as follows.

#### Specification: Print (Child) NFTs

##### **[R6]**

*There MUST be a struct that encodes the necessary data for direct parent-child relationships the minimal set of which is defined below.*
```
///
/// @dev Definition of the struct that allows for direct parent-child relationships.
///
/// @param maxchildren represents the maximum number of children a parent NFT can have.
/// @param ancestrylevel represents the level of ancestry of the child NFTs in a generation from the genesis NFT
/// @param children is the array of children tokenIds of an NFT parent
struct Child {
    uint256 maxchildren;
    uint256 ancestryLevel;
    uint256[] children;

> }

Reviewer 3: Perhaps a dumb question, but I didn't notice `Child` appearing in any public functions/accessors in the spec itself. Does it need to be specified exactly like this, or does an implementation simply have to implement _something_ that enforces the parent-child relationship?

I would assume this also applies to some of the other structs too.

Author: @SamWilsn there are no dumb questions, only dumb answers. The naming does not have to be the same as in the spec but the struct elements listed are required for sure. And yes it does apply to other data structures as well -- naming can be changed, struct elements listed are minimally required.

Reviewer 3: As a thought experiment, would an implementation be able to use a `uint8` for `maxChildren`, if that value was never to go above `255`?

I guess what I'm asking is: is this struct (and the similar ones) _necessary_ to specify in the EIP, or should the EIP specify the intended externally visible behaviour and leave the implementation details up to the implementer?

Author: @SamWilsn sorry overlooked this comment ... i understand where you are coming from. I think one can argue either way ... if you say, the implementer can choose, then how does a user know, or an application know what the max number supported by an implementation is ... I could easily envision a scenario of only 1 but with 10,000 children ... if this is uint256 and an application is spec-compliant, then there will never be an issue. Sooooo ... this is a fine line, IMHO. I tend to fall on the more prescriptive to reduce ambiguity where it can have significant user impact.

```

<a name="r58">**[R58]**</a> *The internal `_royaltyPayOut` function MUST perform the following actions:
* *send the payment to the `payoutaccount`.*
* *update the `royaltybalance` of the `RAsubaccount` of the Royalty Account upon successful transfer.*

<a name="r59">**[R59]**</a> *The following steps MUST be taken to send out a royalty payment to its recipient:*
* *find the Royalty Sub Account.*
* *extract `tokenType` from the Royalty Sub Account.*
* *based on the token type send to the `payoutAccount` either*
    * *'ETH' / relevant protocol token or*
    * *another token based on token type* 
* *and only if the payout transaction is successful, deduct `amount` from `royaltybalance` of the Royalty Sub Account,`RAsubaccount`, and then return `true` as the function return parameter, otherwise return `false`.* 

## Backwards Compatibility
This EIP is backwards compatible to the ERC-721 standard introducing new interfaces and functionality but retaining the core interfaces and functionality of the ERC-721 standard.

## Test Cases
A full test suite is part of the reference implementation.

## Reference Implementation

> The Treetrunk reference implementation of the standard can be found under https://github.com/ethereum/EIPs/tree/master -> assets -> eip-4910 -> eip-4910-reference-implementation.

Reviewer: Please link to the implementation using the relative link format `../assets/eip-4910/etc`.

Author: @lightclient according to @MicahZoltu I would have to link to every single file in the reference implementation (not feasible due to large number) since folder referencing is not supported in this repo. Please, advise!

Reviewer: @MicahZoltu could you elaborate on why a directory listing isn't supported? Is it because the reference implementation should really only be a file or two?

Reviewer: Also is an apache 2.0 license okay for a reference implementation?

Reviewer 2: > @MicahZoltu could you elaborate on why a directory listing isn't supported? Is it because the reference implementation should really only be a file or two?

Jekyll (as currently deployed) just doesn't support it.  You'll get a 404.  I'm not totally against adding directory listing support to Jekyll, but we should keep in mind that this would add an additional constraint on any future rendering systems we might want to migrate to in the future.  I generally would rather avoid the constraint and just require people "manually" handle directory listing (e.g., in the EIP include a list of links, or create a single TOC markdown file in assets directory for the EIP in question.

Reviewer 2: > Also is an apache 2.0 license okay for a reference implementation?

No.  Everything in the EIPs repository should be CC0.  If you can't do CC0 (e.g., because it isn't your code), then you should not include it as a reference implementation (just leave out the section).

Reviewer 3: > No. Everything in the EIPs repository should be CC0. If you can't do CC0 (e.g., because it isn't your code), then you should not include it as a reference implementation (just leave out the section).

I'd pretty strongly disagree with this. EIPs themselves should be CC0, but I see no reason to disallow apache/MIT/BSD/Unlicense/etc. They're freely distributable and impose few restrictions on downstream code. Obviously MPL/GPL/AGPL/LGPL/etc would be bad for reference implementations, however.

Reviewer 2: We should probably discuss this among the editors (either in a separate issue or on the fortnightly call).

Reviewer 3: @poojaranjan can we put this on the next agenda?

Author: > > @MicahZoltu could you elaborate on why a directory listing isn't supported? Is it because the reference implementation should really only be a file or two?
> 
> Jekyll (as currently deployed) just doesn't support it. You'll get a 404. I'm not totally against adding directory listing support to Jekyll, but we should keep in mind that this would add an additional constraint on any future rendering systems we might want to migrate to in the future. I generally would rather avoid the constraint and just require people "manually" handle directory listing (e.g., in the EIP include a list of links, or create a single TOC markdown file in assets directory for the EIP in question.

@MicahZoltu @lightclient @SamWilsn It might be worth looking into how the Decentralized Identity Foundation, or the W3C, or OASIS handle reference implementations. Generally speaking, reference implementations should come with test fixtures and the ability to generate test reports to show compliance with a standard, at least this is how DIF and W3C handle it. Also, reference implementations are often put into a separate repo that the standard links to. That might be a path forward that could satisfy anyone. Happy to support such an effort.

cc @humbitious

Reviewer 2: If we had more manpower, I would probably agree with improving on the reference implementation constraints.  A big factor in EIP editing decision making is "what requires the least amount of effort" because if effort grows, eventually the whole system grinds to a halt.

Author: removed reference implementation

* *update the `royaltybalance` of the `RAsubaccount` of the Royalty Account upon successful transfer.*

<a name="r59">**[R59]**</a> *The following steps MUST be taken to send out a royalty payment to its recipient:*
* *find the Royalty Sub Account.*
* *extract `tokenType` from the Royalty Sub Account.*
* *based on the token type send to the `payoutAccount` either*
    * *'ETH' / relevant protocol token or*
    * *another token based on token type* 
* *and only if the payout transaction is successful, deduct `amount` from `royaltybalance` of the Royalty Sub Account,`RAsubaccount`, and then return `true` as the function return parameter, otherwise return `false`.* 

## Rationale

Royalties for NFTs is at its core a distribution licensing problem. A buyer obtains the right to an asset/content which might or might not be reproducible, alterable etc. by the buyer or agents of the buyer. Therefore, a comprehensive specification must address a hierarchy of royalties, where one or more assets are derived from an original asset as described in the Motivation section in detail. Consequently, a design must solve for a multi-level inheritance, and thus, recursion problem. 

> In order to solve for the complicated inheritance problem, this proposal design breaks down the recursive problem of the hierarchy first into a tree of depth N. And the further breaks down the tree structure into N separate problems, one for each layer. This design allows one to traverse the tree from its lowest level upwards to its root most efficiently. This is achieved with the design for the `distributePayment` function and the NFT data structures allowing for the tree structure e.g. `ancestry',`royaltyAccount`, 'RAsubaccount'. 

Reviewer 4: ```suggestion
In order to solve for the complicated inheritance problem, this proposal design breaks down the recursive problem of the hierarchy first into a tree of depth N. And the further breaks down the tree structure into N separate problems, one for each layer. This design allows one to traverse the tree from its lowest level upwards to its root most efficiently. This is achieved with the design for the `distributePayment` function and the NFT data structures allowing for the tree structure e.g. `ancestry`,`royaltyAccount`, `RAsubaccount`. 
```

```

<a name="r61">**[R61]**</a> *The internal `_royaltyPayOut` function MUST perform the following actions:
* *send the payment to the `payoutaccount`.*
* *update the `royaltybalance` of the `RAsubaccount` of the Royalty Account upon successful transfer.*

<a name="r62">**[R62]**</a> *The following steps MUST be taken to send out a royalty payment to its recipient:*
* *find the Royalty Sub Account.*
* *extract `tokentype` from the Royalty Sub Account.*
* *based on the token type send to the `payoutAccount` either*
    * *'ETH' / relevant protocol token or*
    * *another token based on token type* 
* *and only if the payout transaction is successful, deduct `amount` from `royaltybalance` of the Royalty Sub Account,`RAsubaccount`, and then return `true` as the function return parameter, otherwise return `false`.* 

## Backwards Compatibility
This EIP is backwards compatible to the ERC721 standard introducing new interfaces and functionality but retaining the core interfaces and functionality of the ERC-721 standard.

## Test Cases
A full test suite is part of the reference implementation.

## Reference Implementation

> The Treetrunk reference implementation of the standard can be found [here](https://github.com/ethereum/EIPs/tree/master/assets/eip-4910/eip-4910-reference-implementation/).

Reviewer 2: This link won't work as directory listing is not supported.  You need to link to each individual file from here.

Author: @MicahZoltu ok ... will not link directly.

Reviewer: Could you elaborate on why a directory listing isn't supported?

Author: removed reference implementation

    - NFT phishing attack targeting the `addListNFT` function to trick victim to list one or more NFTs at a very low price and the attacker immediately registering a payment, and executing that payment right away. Mitigation: Implement a waiting period for a purchase can be affected giving the victim time to call the `removeListNFT` function. In addition, an implementer could require Two-Factor-Authentication either built into the contract or by utilizing an authenticator app such as Google Authenticator built into a wallet software. 

Besides the usage of professional security analysis tools, it is also recommended that each implementation performs a security audit of its implementation.

## Copyright

> Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/).

Reviewer 4: ```suggestion
Copyright and related rights waived via [CC0](../LICENSE).
```

#### Paying out Royalties to the NFT Owner -- `from` address in `safeTransferFrom` Function

This is the final part of the proposal.

There are two versions of the payout function -- a `public` function and an `internal` function.

The public function has the following interface:
```
function royaltyPayOut (uint256 tokenId, address RAsubaccount, address payable payoutAccount, uint256 amount) public virtual nonReentrant returns (bool)
```

where we only need the `tokenId`, the RA Sub Account address, `_RAsubaccount` which is the `owner`, and the amount to be paid out, `_amount`. Note that the function has `nonReentrant` modifier protection, because funds are being payed out.

#### Sending a Payout Payment

The following steps need to be taken:
* find the RA Sub Account based on `RAaccount` and the `subaccountPos` and extract the balance
* extract `tokentype` from the Sub Account
* based on the token type, send the payout payment (not exceeding the available balance)

### Data Structures

> #### Rationale: Royalty Account and Royalty Sub Accounts

Reviewer 3: Again, sorry for this taking a couple rounds to convey. I'm still new at this EIP editing thing :sweat_smile: 

The current structure is:

```markdown
# Specification

## Data Structures

### Rationale: Royalty Account and Royalty Sub Accounts

### Specification: Royalty Account and Royalty Sub Accounts
```

It needs to be:

```markdown
# Specification

## Data Structures

### Royalty Account and Royalty Sub Accounts

# Rationale

## Data Structures

### Royalty Account and Royalty Sub Accounts
```

(alternatively, if the rationale is small enough, you can sneak it in the specification section without a header :smirk:)

Author: @SamWilsn  No issue whatsoever. So sorry for the oversight. I overlooked that one area. I snuck all of the rationales in at the beginning of each specification section since they are generally very short, as you suggested.

Reviewer 2: Minor correction, the top level sections should be `##` not `#`.  Don't ask me why, it is one of those things where the original reasoning has been lost to time and we continue doing it because consistency is more important than doing it more correctly.

Author: I doublechecked @SamWilsn and @MicahZoltu and the top-level is `##` in the doc.
